File: java/hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java
package org.acme.schooltimetabling.solver;

import static org.acme.schooltimetabling.TimetableApp.generateDemoData;
import static org.assertj.core.api.Assertions.assertThat;

import java.time.Duration;

import ai.timefold.solver.core.api.solver.Solver;
import ai.timefold.solver.core.api.solver.SolverFactory;
import ai.timefold.solver.core.config.solver.EnvironmentMode;
import ai.timefold.solver.core.config.solver.SolverConfig;

import org.acme.schooltimetabling.TimetableApp;
import org.acme.schooltimetabling.domain.Lesson;
import org.acme.schooltimetabling.domain.Timetable;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfSystemProperty;

@EnabledIfSystemProperty(named = "slowly", matches = "true")
class TimetableEnvironmentTest {

    @Test
    void solveFullAssert() {
        solve(EnvironmentMode.FULL_ASSERT);
    }

    @Test
    void solveStepAssert() {
        solve(EnvironmentMode.STEP_ASSERT);
    }

    void solve(EnvironmentMode environmentMode) {
        SolverFactory<Timetable> solverFactory = SolverFactory.create(new SolverConfig()
                .withSolutionClass(Timetable.class)
                .withEntityClasses(Lesson.class)
                .withConstraintProviderClass(TimetableConstraintProvider.class)
                .withEnvironmentMode(environmentMode)
                .withTerminationSpentLimit(Duration.ofSeconds(30)));

        // Load the problem
        Timetable problem = generateDemoData(TimetableApp.DemoData.SMALL);

        // Solve the problem
        Solver<Timetable> solver = solverFactory.buildSolver();
        Timetable solution = solver.solve(problem);
        assertThat(solution.getScore()).isNotNull();
    }

}

File: getting-started/quarkus-integration-kotlin/src/test/kotlin/org/acme/kotlin/schooltimetabling/rest/TimetableEnvironmentTest.kt
package org.acme.kotlin.schooltimetabling.rest

import ai.timefold.solver.core.api.solver.Solver
import ai.timefold.solver.core.api.solver.SolverFactory
import ai.timefold.solver.core.config.solver.EnvironmentMode
import ai.timefold.solver.core.config.solver.SolverConfig
import io.quarkus.test.junit.QuarkusTest
import io.restassured.RestAssured.given
import jakarta.inject.Inject
import org.acme.kotlin.schooltimetabling.domain.Timetable
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.condition.EnabledIfSystemProperty
import java.time.Duration

@QuarkusTest
@EnabledIfSystemProperty(named = "slowly", matches = "true")
class TimetableEnvironmentTest {

    @Inject
    var solverConfig: SolverConfig? = null

    @Test
    fun solveFullAssert() {
        solve(EnvironmentMode.FULL_ASSERT)
    }

    @Test
    fun solveStepAssert() {
        solve(EnvironmentMode.STEP_ASSERT)
    }

    fun solve(environmentMode: EnvironmentMode) {
        // Load the problem
        val problem: Timetable = given()
            .`when`()["/demo-data/SMALL"]
            .then()
            .statusCode(200)
            .extract()
            .`as`(Timetable::class.java)

        // Update the environment
        val updatedConfig = solverConfig!!.copyConfig()
            .withTerminationSpentLimit(Duration.ofSeconds(30))
            .withEnvironmentMode(environmentMode)
        val solverFactory: SolverFactory<Timetable> = SolverFactory.create(updatedConfig)

        // Solve the problem
        val solver: Solver<Timetable> = solverFactory.buildSolver()
        val solution: Timetable = solver.solve(problem)
        assertThat(solution.score?.isFeasible).isTrue()
    }
}
